execution_engine/src/storage/trie_store/operations/mod.rs

```diff
@@ -505,15 +512,127 @@ where
+                // hit this.
+                unreachable!()
+            }
+            Trie::Node { pointer_block } => {
+                let index = {
+                    assert!(depth < path.len(), "depth must be < {}", path.len());
+                    path[depth]
+                };
+                let maybe_pointer: Option<Pointer> = {
+                    let index: usize = index.into();
+                    assert!(index < RADIX, "index must be < {}", RADIX);
+                    pointer_block[index]
+                };
+                let pointer = match maybe_pointer {
+                    Some(pointer) => pointer,
+                    None => {
+                        return Ok(TrieScanRaw::new(
+                            Either::Right(Trie::Node { pointer_block }),
+                            acc,
+                        ));
+                    }
+                };
+                match store.get_raw(txn, pointer.hash())? {
+                    Some(next) => {
+                        current = next;
+                        depth += 1;
+                        acc.push((index, Trie::Node { pointer_block }))
+                    }
+                    None => {
+                        panic!(
+                            "No trie value at key: {:?} (reading from path: {:?})",
+                            pointer.hash(),
+                            path
+                        );
+                    }
+                }
+            }
+            Trie::Extension { affix, pointer } => {
+                let sub_path = &path[depth..depth + affix.len()];
+                if sub_path != affix.as_slice() {
+                    return Ok(TrieScanRaw::new(
+                        Either::Right(Trie::Extension { affix, pointer }),
+                        acc,
+                    ));
+                }
+                match store.get_raw(txn, pointer.hash())? {
+                    Some(next) => {
+                        let index = {
+                            assert!(depth < path.len(), "depth must be < {}", path.len());
+                            path[depth]
+                        };
+                        current = next;
+                        depth += affix.len();
+                        acc.push((index, Trie::Extension { affix, pointer }))
+                    }
+                    None => {
+                        panic!(
+                            "No trie value at key: {:?} (reading from path: {:?})",
+                            pointer.hash(),
+                            path
+                        );
+                    }
+                }
+            }
+        }
+    }
+}
+
 #[derive(Debug, PartialEq, Eq)]
 pub enum DeleteResult {
     Deleted(Digest),
     DoesNotExist,
     RootNotFound,
 }
 
-#[allow(unused)]
-fn delete<K, V, T, S, E>(
+/// Delete provided key from a global state so it is not reachable from a resulting state root hash.
+pub fn delete<K, V, T, S, E>(
```

Comment:
```suggestion
pub(crate) fn delete<K, V, T, S, E>(
```

Reply:
Did this - it's unfortunate DeleteResult was pub all the time though but we'll deal with it in #3828